### PR TITLE
Move to Claude

### PR DESCRIPTION
**Pull request by Nokondi**
## Description
Updates aiproxy to use Claude 3.5 Sonnet as our default model. Includes updates to LLM response validation and processing. Includes a temporary script to upload new releases of the aitt_release_data to AWS S3.

upload_new_release_to_s3.py terminal output:
image

Folder in S3:
image
image

## Links
[aitt_release_data Claude PR](https://github.com/code-dot-org/aitt_release_data/pull/1)
## Testing story
Manually tested the upload script. Other fixes did not need additional unit tests.
## Accuracy
The Claude update will negatively impact accuracy on some lessons compared to our current release using GPT 4 turbo, but overall accuracy is unaffected.
| Lesson | Current Release | New Release |
|--------|--------|--------|
| csd3-2023-L11 | 84% | 75% |
| csd3-2023-L14 | 86% | 80% |
| csd3-2023-L18 | 75% | 69% |
| csd3-2023-L21 | 73% | 79% |
| csd3-2023-L24 | 76% | 86% | 
| csd3-2023-L28 | 65% | 70% | 
| **average across all lessons** | **77%** | **77%** |
## API Changes
N/A
## Follow-up work
Replace the release upload script with a CICD pipeline to automatically upload new releases to S3
## PR Checklist:
- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] Relevant documentation has been added or updated
- [ ] Accuracy against the dev set has been retested
- [ ] Changes in accuracy have been communicated
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: bin/upload_new_release_to_s3.py
Comment (Nokondi) [inline bin/upload_new_release_to_s3.py:47]: I put this here instead of in the aitt_release_data repo to avoid having to install another instance of aws_google.

File: cicd/1-setup/cicd-dependencies.template.yml
Comment (cearachew) [inline cicd/1-setup/cicd-dependencies.template.yml:157]: Is this for the future CICD work, or is this needed for the upload to s3 script?

File: bin/upload_new_release_to_s3.py
Comment (cearachew) [inline bin/upload_new_release_to_s3.py:12]: If we're planning to switch to an automatic CICD release on `aitt_release_data` soon it's probably nbd, but I think a standardized release name with a date would be better than manually naming the release

File: lib/assessment/label.py
Comment (cearachew) [inline lib/assessment/label.py:412]: Why replace the straight quotes with smart quotes?

File: bin/upload_new_release_to_s3.py
Comment (Nokondi) [reply bin/upload_new_release_to_s3.py:12]: Agreed. I believe what Dave and I had discussed was using the branch name and date to name the release, which shouldn't be too hard to implement. I'm going to chat with infra about getting that set up sooner than later.

File: lib/assessment/label.py
Comment (Nokondi) [reply lib/assessment/label.py:412]: Claude returns canceled straight quotes in Key Concepts that have quotes, which causes a mismatch, because we use smart quotes. I tried changing it the other ways (swap out our smart quotes with straight quotes) but this was a quicker fix, and it doesn't cause issues with other models.

Comment (cearachew) [reply lib/assessment/label.py:412]: That makes sense. Imo stripping out all quotes and whitespace from the key concepts/ scores would be better, I guess I'm not sure why there are quotes in those strings at all? But also, 🤷‍♀️ a quick fix here seems fine

**Review by cearachew — APPROVED** [review]
LGTM, thanks!!

File: bin/upload_new_release_to_s3.py
Comment (davidsbailey) [reply bin/upload_new_release_to_s3.py:12]: it sounds like we've had a few pairwise conversations on this and I think I might be out of the loop on some discussion too. I was thinking that it would be ideal if we could use the release name to quickly find the corresponding contents in github, and the first thing that comes to mind is to have the git branch name match the release name. What do you think about:
1. have the aitt_release_data branch be in the format of YYYY-MM-DD-your-description-here
2. have this script verify the aitt branch is in that format, and then use it as the release name

side note, I made sure that "auto delete branches for merged PRs" is disabled in the aitt_release_data repo, and if we like this strategy we should try to remember not to delete branches after merging PRs in that repo.

Comment (davidsbailey) [reply bin/upload_new_release_to_s3.py:12]: > I think a standardized release name with a date would be better than manually naming the release

not sure if my comments are agreeing or disagreeing with this statement Ceara 😅 I'd also be fine with just using a datestamp if that sounds better to you, though in that case I think we need somewhere in the aitt repo (a release notes file?) that we could look to see a name / description for what changed in that release.

File: lib/assessment/label.py
Comment (davidsbailey) [reply lib/assessment/label.py:412]: > Claude returns canceled straight quotes in Key Concepts that have quotes

@Nokondi can you please give an example showing what you mean by this?

> That makes sense. Imo stripping out all quotes and whitespace from the key concepts/ scores would be better, I guess I'm not sure why there are quotes in those strings at all? But also, 🤷‍♀️ a quick fix here seems fine

@cearachew I agree here also. please correct my understanding but this would mean modifying the system prompt until the AI model(s) no longer returned key concepts in quotes, is that correct? perhaps this is something we could put on a list to do "before we expand our pool of lessons".

File: lib/assessment/rubric_tester.py
Comment (davidsbailey) [inline lib/assessment/rubric_tester.py:369]: remove debug code?

File: bin/upload_new_release_to_s3.py
Comment (cearachew) [reply bin/upload_new_release_to_s3.py:12]: I think we're in agreement generally (release names should have a name that makes it easy to see what the latest is and it should be easy to find the corresponding contents in git) and we are figuring out the easiest way to do so

Comment (cearachew) [reply bin/upload_new_release_to_s3.py:12]: IMO we should `git tag` each merge to main in `aitt_release_data` with either the date or major.minor.patch and make a tag release w/ CICD to s3. Then you could just checkout the tag in git to get the content

Comment (cearachew) [reply bin/upload_new_release_to_s3.py:12]: i.e. we should do this: https://github.com/code-dot-org/code-dot-org/releases

File: lib/assessment/label.py
Comment (cearachew) [reply lib/assessment/label.py:412]: I don't think this means modifying the prompts at all. It just means post-processing to normalize the model outputs and the comparison strings from the rubrics to strip surrounding whitespace/quotes/whatever

Comment (davidsbailey) [reply lib/assessment/label.py:412]: ok, yes, I see what you're saying and think it sounds doable without waiting for lesson expansion, though also agree it doesn't seem like a blocker for this PR

File: bin/upload_new_release_to_s3.py
Comment (davidsbailey) [reply bin/upload_new_release_to_s3.py:12]: 💯 on tags! Mark, do you want to choose between making the branch name match the release name, or switching this to use tags instead? the tags change initially seemed like a slightly bigger amount of effort that I didn't want to block the claude move, but (now that Karim is unblocked via adhoc) it seems reasonable to do as follow up if there is sufficient motivation

File: lib/assessment/label.py
Comment (Nokondi) [reply lib/assessment/label.py:412]: For whatever reason, our rubrics and expected grades files use smart quotes in some key concepts, so if the LLM response doesn't return identical key concepts, you end up with a error. Doesn't seem to have been an issue with GPT, but Claude only returns straight quotes. I'm fine with removing the quotes as an alternate solution.

Comment (davidsbailey) [reply lib/assessment/label.py:412]: I see now we are talking about quotes within some Key Concept names, rather than surrounding them:
* https://github.com/code-dot-org/aitt_release_data/blob/758ea5a226bada8f457e14894d1750b60c225cf2/csd3-2023-L14/standard_rubric.csv#L4
* 
```
(.venv) Dave-MBP:~/src/aiproxy/datasets/contractor-grades-batch-1-fall-2023 (main)$ cat csd3-2023-L14/actual_labels.csv
student,Modularity - Sprites and Sprite Properties,Position and Movement,Optional “Stretch” Feature - Variables
...
```
normalizing to curly quotes in this way seems a bit fragile because it assumes there is exactly 0 or 1 pair of curly quotes within the string and that there are no straight quotes. OTOH if we normalize to anything else (straight quotes or no quotes) it will be slightly more work because we'll also have to normalize what we read out of the standard_rubric.csv and actual_labels.csv files. 

since we'll get pretty immediate validation error if we introduce straight quotes into a key concept in the future, and this prob won't happen until we expand to more lessons, I think I feel good about shipping what you have here Mark, and opening a work item to instead normalize to straight quotes as part of our "before we expand to more lessons" list of work items. 

how does this sound to you all?

Comment (cearachew) [reply lib/assessment/label.py:412]: I think that sounds good, I agree that normalizing to replace all quotes with straight quotes for comparison is less fragile than the other way around

Comment (Nokondi) [reply lib/assessment/label.py:412]: Sounds fine to me. Yeah, I realized it's a hacky solution, but felt safe for now, since we only have one pair of quotes per Key Concept. I think the easiest fix down the road is to just remove punctuation from the Key Concepts in the rubric and labels files.

**Comment by cat5inthecradle** [comment]
I'm wondering a couple of things, no good way to make these inline comments.

1. Should the auth requirement apply to all routes?
2. Should we use the [Authorization](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization) header?
3. Should we return a proper [401](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/401) HTTP Code?

The implementation of all three of those suggestions might be possible by adding the following to the `create_app` function in "src/__init__.py"

```
@app.before_request
def require_api_key():
    api_key = request.headers.get('Authorization')
    if api_key != os.getenv('API_KEY'):
        return jsonify({"error": "Unauthorized"}), 401
```

File: bin/upload_new_release_to_s3.py
Comment (cat5inthecradle) [reply bin/upload_new_release_to_s3.py:12]: My vote would be tags or commit hashes. Tags make a lot of sense, given that these are indeed releases.

File: cicd/1-setup/cicd-dependencies.template.yml
Comment (cat5inthecradle) [reply cicd/1-setup/cicd-dependencies.template.yml:157]: I'm tracking changes to this template and applying them manually to production as we iterate on them. The alternative would be to pop these out into their own PR and merge them ahead of this to get them working, that's just a little more work, and it's nice to see it all in this PR.

**Comment by davidsbailey** [comment]
> I'm wondering a couple of things, no good way to make these inline comments.
> 
> 1. Should the auth requirement apply to all routes?
> 2. Should we use the [Authorization](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization) header?
> 3. Should we return a proper [401](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/401) HTTP Code?
> 
> The implementation of all three of those suggestions might be possible by adding the following to the `create_app` function in "src/__init__.py"
> 
> ```
> @app.before_request
> def require_api_key():
>     api_key = request.headers.get('Authorization')
>     if api_key != os.getenv('API_KEY'):
>         return jsonify({"error": "Unauthorized"}), 401
> ```

This makes sense to me! Nit: throughout this snippet, let's call it something more specific, like AIPROXY_API_KEY, to disambiguate it from the openai API key or maybe an AWS key

**Comment by cat5inthecradle** [comment]
Good point, I was thinking AIProxy was redundant, but was totally forgetting about the openai key.

File: docker-compose.yml
Comment (Nokondi) [inline docker-compose.yml:1]: This was an automatic linting change from VS Code

File: bin/upload_new_release_to_s3.py
Comment (cat5inthecradle) [reply bin/upload_new_release_to_s3.py:12]: In the interest of resolving this conversation - is name good for now or does this need to change to advise something more specific (like version number or date)

File: src/assessment.py
Comment (wilkie) [inline src/assessment.py:27]:
```suggestion

```

File: bin/upload_new_release_to_s3.py
Comment (wilkie) [inline bin/upload_new_release_to_s3.py:32]: does `README.md` always exist when `.git` does? this would technically crash if not.

You could import:

```
from contextlib import suppress
```

And then:

```
files = os.listdir(full_path)
with suppress(ValueError):
    files.remove('.git')
    files.remove('README.md')
for f in files:
...
```

To just try to remove it and ignore the errors when it isn't found.

Comment (Nokondi) [reply bin/upload_new_release_to_s3.py:32]: This isn't removing them from the repo, it's just removing them from the file list that's being parsed for the upload to S3. We don't need to upload the Readme and .git folder.

Comment (Nokondi) [reply bin/upload_new_release_to_s3.py:32]: This should only return true at the root of the aitt_release_data repo, and it's a temp solution while we get cicd setup. I can add the suppress error code, but I don't think it's necessary, since this script has such a specific stopgap use.

File: bin/upload_new_release_to_s3.py
Comment (davidsbailey) [reply bin/upload_new_release_to_s3.py:12]: making the branch name match the date-stamped release name also works fine for me if that's any easier, and we'll just keep in mind that before a certain date, releases were branches and not tags.

File: lib/assessment/label.py
Comment (davidsbailey) [inline lib/assessment/label.py:144]: nice, I like that this is more targeted!

File: src/assessment.py
Comment (davidsbailey) [inline src/assessment.py:66]: unfortunately we depend on "openai" being present in this string, here: https://github.com/code-dot-org/code-dot-org/blob/0b263e3f3957bade5036eb8d54f98a06e2bda876/dashboard/app/jobs/evaluate_rubric_job.rb#L193-L207

before merging this PR, please make sure dashboard is updated to accept both strings. matching on "LLM timeout" rather than just "LLM".

File: bin/upload_new_release_to_s3.py
Comment (davidsbailey) [inline bin/upload_new_release_to_s3.py:1]: since we don't typically write tests for bin scripts, can you please show some command line output showing that this is working? for example:
* run the script, uploading to temporary / throwaway release name starting with `tmp-` or similar
* make sure we can see which version of this bin script you ran as well as what point in the aitt_release_data repo you ran against

**Comment by cat5inthecradle** [comment]
I extracted some Infra changes into their own PR. If we merge #88 first, then we can get all secrets situated prior to merging this code that will depend on them.

File: lib/assessment/label.py
Comment (Nokondi) [reply lib/assessment/label.py:114]: It was just a guess. What would you recommend as a max value?

Comment (davidsbailey) [reply lib/assessment/label.py:114]: TLDR: I vote for removing this line.

upon further inspection, it doesn't look like `max_tokens` is in the request API: https://docs.anthropic.com/en/api/complete there is a `max_tokens_to_sample`. 

according to https://docs.anthropic.com/en/docs/about-claude/models, the default limit is 4096 tokens, which can be increased to 8192 tokens by using a special header. 

so, let's just use the default limit, and if we notice that we are getting a lot of responses rejected due to hitting token limits before 8192 becomes the default, we can consider using the header to increase the limit.

I did check the reports in aitt_release_data for the latest release, and I didn't see any errors there, which is encouraging.

**Review by davidsbailey — COMMENTED** [review]
@Nokondi I saw you re-requested review. I have some outstanding comments from the previous review, please lmk if it would be helpful for me to go through and reply to the ones that looks like they haven't been addressed yet.

**Comment by Nokondi** [comment]
> @Nokondi I saw you re-requested review. I have some outstanding comments from the previous review, please lmk if it would be helpful for me to go through and reply to the ones that looks like they haven't been addressed yet.

OK, I'll double-check what I missed.

File: src/assessment.py
Comment (Nokondi) [reply src/assessment.py]: I updated this to return either openai or bedrock depending on the set llm_model value. I also updated cdo to match on either value in the code you linked in this PR: https://github.com/code-dot-org/code-dot-org/pull/60063

File: bin/upload_new_release_to_s3.py
Comment (davidsbailey) [reply bin/upload_new_release_to_s3.py:12]: > let's keep our releases using the YYYY-MM-DD- prefix for now, which rules out using commit hashes as the release names. let's plan on tags as the long-term solution.
> 
> knowing that, how about once this is merged, you manually make a tag which matches the name of the date-prefixed release? this would support the goal of making it easy to locate the release contents in github given the release name, while building in the right direction.

@Nokondi can you confirm if this sounds good?

File: bin/upload_new_release_to_s3.py
Comment (davidsbailey) [reply bin/upload_new_release_to_s3.py:1]: @Nokondi can you please show how you validated this script, and provide the info listed above?

**Review by davidsbailey — COMMENTED** [review]
Hey Mark, I think what's missing for me is to understand what your plan is to get this all merged. If you merge this as-is ahead of the CDO PR, it seems like TA will break in production because it will not yet be providing the new auth key. if you try to merge both PRs at the same time, that might work but could turn into a bit of a mess if we need to roll back. 

if you already have a list of steps in mind that'd be great to see written down, otherwise lets use checkin time to work out a plan for getting this merged and deployed.

File: bin/upload_new_release_to_s3.py
Comment (Nokondi) [reply bin/upload_new_release_to_s3.py:1]: I posted the script output in the PR comments. Will update with additional info.

File: bin/upload_new_release_to_s3.py
Comment (Nokondi) [inline bin/upload_new_release_to_s3.py]: This was the version of the script that I ran to generate the output in the PR comments. It posted the aitt_release_data contents to teaching_assistant/experiments/2024-08-07-claude-test in the cdo-ai S3 bucket.

**Comment by Nokondi** [comment]
> Hey Mark, I think what's missing for me is to understand what your plan is to get this all merged. If you merge this as-is ahead of the CDO PR, it seems like TA will break in production because it will not yet be providing the new auth key. if you try to merge both PRs at the same time, that might work but could turn into a bit of a mess if we need to roll back.
> 
> if you already have a list of steps in mind that'd be great to see written down, otherwise lets use checkin time to work out a plan for getting this merged and deployed.

Plan is as follows:
1. CDO security-only update: This will set the CDO server to start sending the aiproxy api key for authorization (which aiproxy will ignore initially). Will get this set up now that the new adhoc branch is created.
2. aiproxy update (this PR): This will set the security on aiproxy so it will now read the api key and block any other traffic. Also adds code to process claude responses and other necessities for the switch
3. aitt_release_data: Will push this to S3 releases directory since the PR has already been approved
4. CDO update to point to new release

File: bin/upload_new_release_to_s3.py
Comment (davidsbailey) [reply bin/upload_new_release_to_s3.py:1]: thank you, apologies, i didn't notice the update to the PR description. A couple of issues:
1. I don't see the new release listed in the PR description in S3. can you please share the full URL where you're seeing what you showed in the screenshot? I checked https://us-east-1.console.aws.amazon.com/s3/buckets/cdo-ai?region=us-east-1&bucketType=general&prefix=teaching_assistant/releases/&showversions=false as well as using the command line:
```
$ aws s3 ls s3://cdo-ai/teaching_assistant/releases/
                           PRE 2024-02-01-ai-rubrics-pilot-baseline/
                           PRE 2024-02-07-ai-rubrics-pilot-line-annotations/
                           PRE 2024-02-28-ai-rubrics-json-reason/
                           PRE 2024-03-06-ai-rubrics-json-reason-confidence-exact/
                           PRE 2024-03-11-ai-rubrics-json-evidence/
                           PRE 2024-03-15-ai-rubrics-json-evidence/
                           PRE 2024-03-25-confidence-exact/
                           PRE 2024-04-26-confidence-autogen/
                           PRE 2024-04-30-confidence-autogen-turbo/
                           PRE 2024-07-10-cfe-release/
```
2. in part of the output, the release name is YYYY-MM-DD-test, and in another part it is test-YYYY-MM-DD. I assume we want the date prefix in both parts of the output

File: src/assessment.py
Comment (davidsbailey) [inline src/assessment.py:70]: I don't think we want to silently drop this error in other cases, for example if we try out another LLM (or if we switch to accessing AWS models without bedrock). I would suggest adding a final else clause:
```
        else:
            return f"LLM timeout: #{e}: ", 504
```

**Comment by davidsbailey** [comment]
> Plan is as follows:
> 
> 1. CDO security-only update: This will set the CDO server to start sending the aiproxy api key for authorization (which aiproxy will ignore initially). Will get this set up now that the new adhoc branch is created.
> 2. aiproxy update (this PR): This will set the security on aiproxy so it will now read the api key and block any other traffic. Also adds code to process claude responses and other necessities for the switch
> 3. aitt_release_data: Will push this to S3 releases directory since the PR has already been approved
> 4. CDO update to point to new release

this ordering sounds great! thank you for clarifying. I looked through https://github.com/code-dot-org/code-dot-org/pull/60063 and it looks backward-compatible, once the two sections marked as adhoc-specific are removed.

File: lib/assessment/label.py
Comment (davidsbailey) [inline lib/assessment/label.py:129]: per the notes in the jira ticket, the timeout here should be 150 seconds. this can be done in a next PR if needed, but should be done before we launch the claude release. some reasoning here: https://github.com/code-dot-org/code-dot-org/pull/58411 --> "timeouts" section in PR description

**Comment by davidsbailey** [comment]
this is shaping up nicely Mark! There is one other topic from the jira ticket that I'd like to discuss before we launch:
* add unit tests for bedrock codepaths in aiproxy. 
  * what coverage do we have today for the bedrock codepaths? 
  * do we want to move the bulk of the existing unit tests over from openai to bedrock, and if so, when?

I don't think this should block this PR, but we should make a decision on this before launch.

File: lib/assessment/label.py
Comment (cat5inthecradle) [inline lib/assessment/label.py:398]: Non blocking, but I'm picking up a code smell here that makes me a little nervous about the future maintainability of this code if we need to add/switch models frequently. It seems like we're choosing our model at the beginning of the flow, but then 'forgetting' what model we're using and examining the responses to infer the model and how to parse it's responses.

I wonder if it would be beneficial to add an abstraction layer here to shove model/provider specific code into dedicated classes that accept the same inputs and return normalized outputs. The choice is between having some code duplication across various classes or scattering conditionals throughout the code - hard to say what the best way is right now. If we start doing shenanigans like `if looks_like('model4') || is_kinda('model3') then response.tolowercase` then it might be time to refactor.

File: lib/assessment/config.py
Comment (Nokondi) [reply lib/assessment/config.py:5]: I think the comment refers to the general gpt-4 model tag. We use the specific release tags instead.

File: lib/assessment/label.py
Comment (Nokondi) [reply lib/assessment/label.py]: Turns out max_tokens is required. Messages to Claude 3 sent without it fail.
`botocore.errorfactory.ValidationException: An error occurred (ValidationException) when calling the InvokeModel operation: Malformed input request: #: subject must not be valid against schema 
{
    "required": [
        "messages"
    ]
}
#: required key [max_tokens] not found, please reformat your input and try again.`

Comment (Nokondi) [reply lib/assessment/label.py]: I'm adding it back in and setting it to the default 4096 value. If we get errors, I'll bump it up to 8192.

Comment (davidsbailey) [reply lib/assessment/label.py]: sounds good. for full disclosure, my intuition on this may have been off because I was thinking of the limits we ran into with context window (including both inputs and outputs). I don't believe we've run into any problems with the limit of 4096 output tokens. most other models we've tried have had the same limit on output tokens.

File: lib/assessment/label.py
Comment (davidsbailey) [reply lib/assessment/label.py:399]: I agree with this feedback. one idea is to use an abstract base class to represent the model which various specific LLMs inherit from. in any case, let's consider this as potential follow-up work.

Comment (Nokondi) [reply lib/assessment/label.py:399]: I think that's a good idea. Most of this code was written specifically for connecting with openai, so it would be good to do a refactor. I'll add it as a follow up jira task that we can possibly tackle after the release this month.

File: bin/upload_new_release_to_s3.py
Comment (Nokondi) [reply bin/upload_new_release_to_s3.py:1]: Sorry, I had deleted the folder after getting the screenshots. I'll upload it again so that you can see it.

File: bin/upload_new_release_to_s3.py
Comment (davidsbailey) [reply bin/upload_new_release_to_s3.py:48]: yep, agree all round -- putting it in the aitt repo would be better, but fine to leave it here given that this is an intermediate solution.

File: bin/upload_new_release_to_s3.py
Comment (davidsbailey) [inline bin/upload_new_release_to_s3.py:48]:
```suggestion
print(f"Upload complete! Update S3_AI_RELEASE_PATH at cdo/dashboard/app/jobs/concerns/ai_rubric_config.rb with the new release location: 'teaching_assistant/releases/{now}-{args.release_name}/'")
```

File: bin/upload_new_release_to_s3.py
Comment (davidsbailey) [inline bin/upload_new_release_to_s3.py:15]: to avoid the issue you ran into below, I would recommend introducing a new variable here 
```
release_name = f"{now}-{args.release_name}"
```

File: bin/upload_new_release_to_s3.py
Comment (davidsbailey) [reply bin/upload_new_release_to_s3.py:33]: I think this all exposes the issue that the config we plan to upload should really be isolated to a single directory so that we can recursively copy it without special additional logic. to consider doing in the future

File: bin/upload_new_release_to_s3.py
Comment (davidsbailey) [reply bin/upload_new_release_to_s3.py:1]: this looks more complicated than it needs to be but I am convinced that it is working! 👍 LGTM given that we plan to throw this away when we move to CI/CD.

**Review by davidsbailey — APPROVED** [review]

**Pull request merged**
